src/workers/helpers.js

```diff
@@ -1,53 +1,238 @@
+      extractBuffers(element, store);
+    }
+  } else if (object.constructor == Object) {
+    for (const [key, element] of Object.entries(object)) {
+      extractBuffers(element, store);
+    }
+  } else if (ArrayBuffer.isView(object)) {
+    if (!(object.buffer instanceof ArrayBuffer)) {
+      throw "only ArrayBuffers should be in the message payload";
     }
+    store.push(object.buffer);
+  }
 }
 
 export function postAttempt(step) {
+  postMessage({
+    type: `${step}_START`,
+  });
+}
+
+export function postSuccess(step, info) {
+  if (typeof info == "undefined") {
     postMessage({
-        type: `${step}_START`
+      type: `${step}_CACHE`,
     });
+  } else {
+    var transferable = [];
+    extractBuffers(info, transferable);
+    postMessage(
+      {
+        type: `${step}_DATA`,
+        resp: info,
+      },
+      transferable
+    );
+  }
 }
 
-export function postSuccess(step, info) {
-    if (typeof info == "undefined") {
-        postMessage({
-            type: `${step}_CACHE`
-        });
-    } else {
-        var transferable = [];
-        extractBuffers(info, transferable);
-        postMessage({
-            type: `${step}_DATA`,
-            resp: info
-        }, transferable);
+export function postError(type, err, fatal) {
+  postMessage({
+    type: `${type}_ERROR`,
+    resp: {
+      reason: err.toString(),
+      fatal: fatal,
+    },
+  });
+}
+
+export function splitMetricsByBlock(metrics, blockLevels, blockIds) {
+  var output = {};
+  var blocks = blockIds.array();
+  for (var b = 0; b < blockLevels.length; b++) {
+    let current = {};
+    for (const [key, val] of Object.entries(metrics)) {
+      current[key] = val.array().filter((x, i) => blocks[i] == b);
     }
+    output[blockLevels[b]] = current;
+  }
+  return output;
 }
 
-export function postError(type, err, fatal) {
-    postMessage({
-        type: `${type}_ERROR`,
-        resp: {
-            reason: err.toString(),
-            fatal: fatal
-        },
-    });
+export function splitThresholdsByBlock(thresholds, blockLevels) {
+  var output = {};
+  for (const x of blockLevels) {
+    output[x] = {};
+  }
+
+  for (const [key, val] of Object.entries(thresholds)) {
+    for (var b = 0; b < blockLevels.length; b++) {
+      output[blockLevels[b]][key] = val[b];
+    }
+  }
+
+  return output;
+}
+
+export async function fetchStepSummary(state, step) {
+  // do not send any response to UI if they have not changed
+  if (!state[step].changed) {
+    return null;
+  }
+
+  if (step === "inputs") {
+    let output = {};
+
+    let ngenes = {};
+    for (const a of state[step].fetchCountMatrix().available()) {
+      ngenes[a] = state[step].fetchCountMatrix().get(a).numberOfRows();
+    }
+
+    let gene_info = {};
+    for (const [k, v] of Object.entries(
+      state[step].fetchFeatureAnnotations()
+    )) {
+      let info = {};
+      for (const c of v.columnNames()) {
+        info[c] = bioc.CLONE(v.column(c));
```

Comment:
Might want to check that you're dealing with an `Array`, as nested objects are possible. You probably don't even want TypedArrays in here anyway.

If you are handling an `Array`, you can omit the `CLONE` because plain Arrays don't have buffers that could be transferred anyway.